### DISCUSSION TITLE
Authelia Telemety Settings / Grafana / Prometheus

### DISCUSSION THREAD

**Discussion by TroRon** [authelia/authelia#7315] [Q&A] [answered]
I am currently working on the telemetry section within the Authelia configuration.

I actually want to get the Grafana Dashboard (see Examples) to work.

Grafana / Prometheus are running inside Docker. That works.
But I can't get the Prometheus / Authelia connection to work.

According to the config I added the telemetry block, Authelia also says that it listens on port 9959. But when I check the port it is not open.

Firewall is definitely not.

Do any of you have this setup running and any idea what else it could be?

Authelia Log:
![image](https://github.com/authelia/authelia/assets/62324756/b7d708ca-a9b6-41b4-bb14-098d732d65f4)

Testing:
![image](https://github.com/authelia/authelia/assets/62324756/545d9be4-82c1-4f79-8a1d-e569bc87996e)

Everything runs on the same machine

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

To troubleshoot the issue with Prometheus not being able to connect to Authelia on port 9959, even though Authelia is configured to listen on this port, you can follow these steps:

1. **Verify Authelia Configuration**: Ensure that the telemetry configuration in Authelia is correctly set to enable the Prometheus metrics exporter. The configuration should look something like this:
   ```yaml
   telemetry:
     metrics:
       enabled: true
       address: 'tcp://:9959/'
   ```
   Make sure that `enabled` is set to `true` and the `address` is correctly formatted.

2. **Check Docker Network Settings**: Since both Prometheus and Authelia are running inside Docker, ensure that they are on the same Docker network or can communicate over the network. You might need to check the network settings in your Docker Compose or Docker command.

3. **Inspect Authelia Logs**: Look at the Authelia container logs to see if there are any errors or warnings related to the metrics server starting up or listening on the port. This can give you clues about what might be going wrong.

4. **Test Network Accessibility**: From within the Prometheus container, try to telnet or curl the Authelia metrics endpoint:
   ```bash
   curl http://authelia:9959/metrics
   ```
   This can help verify network connectivity between the Prometheus and Authelia containers.

5. **Review Prometheus Configuration**: Ensure that your Prometheus configuration is set up to scrape metrics from Authelia. The job configuration in Prometheus should look something like this:
   ```yaml
   - job_name: 'authelia'
     static_configs:
       - targets: ['authelia:9959']
   ```
   Replace `'authelia:9959'` with the appropriate address if Authelia is reachable via a different hostname or IP in your setup.

6. **Check for Port Conflicts**: Even though you mentioned the firewall is not an issue, ensure there are no other services running inside or outside Docker that might be using port 9959.

If after these steps the issue persists, it might be helpful to provide more specific logs or error messages from both the Authelia and Prometheus containers to further diagnose the problem.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment, marked as answer]
You've not published the port in docker.

**Reply by TroRon** [reply]
You mean within the Authelia container

9959:9959?

I tried that, same message

![image](https://github.com/authelia/authelia/assets/62324756/99ee2981-52f0-48d2-9d43-8f9fbe37d979)

![image](https://github.com/authelia/authelia/assets/62324756/44aa85e2-8f10-498b-b116-9b158a6a4d16)

**Reply by james-d-elliott** [reply]
Yeah you need to publish the port per standard docker semantics for the iptables rules to be created to translate the port.

**Reply by james-d-elliott** [reply]
`docker exec authelia wget -O - http://127.0.0.1:9959/metrics`

**Reply by TroRon** [reply]
authelia:
    restart: unless-stopped
    image: authelia/authelia
    container_name: authelia_server
    volumes:
      - /docker/authelia:/config
    ports:
      - 19091:9091
      - 9959:9959
    environment:
      - TZ=Europe/Zurich
      
![image](https://github.com/authelia/authelia/assets/62324756/9382f920-1136-420c-8b92-926eab938c36)

**Reply by TroRon** [reply]
URL without /metrics works now:
![image](https://github.com/authelia/authelia/assets/62324756/0d8cb34e-08d9-40f9-a04f-79ed1a009b5d)
![image](https://github.com/authelia/authelia/assets/62324756/d20fc788-8097-4c75-8305-bc31925ac40d)

**Reply by TroRon** [reply]
![image](https://github.com/authelia/authelia/assets/62324756/b73ddbaf-4f10-4adb-8d08-cbf7083714c4)

Solved, add /metrics in the Authelia Config ;)

**Reply by Lebowski89** [reply]
> Yeah you need to publish the port per standard docker semantics for the iptables rules to be created to translate the port.

Not required from what I can see. At least on the same network, at least. See my reply below.

**Comment by TroRon** [comment]
![image](https://github.com/authelia/authelia/assets/62324756/a7e786ad-f5fc-48d8-9b85-e8172c5da482)

A little bit is already displayed ;)

**Comment by Lebowski89** [comment]
Thank you for this topic. 

Per the example configuration (included on this page: https://www.authelia.com/configuration/telemetry/metrics/), the address is:

``` address: 'tcp://:9959/'```

However, doing this will just lead to 404 in Prometheus. It needs to be:  

```address: 'tcp://:9959/metrics'```

This works without any need to expose 9959. Can we get that example config changed, please?

**Reply by james-d-elliott** [reply]
A PR would be welcomed.

> However, doing this will just lead to 404 in Prometheus. It needs to be:

It's important to note the scrape path in Prometheus is configurable. So this statement is not strictly accurate.

> This works without any need to expose 9959. 

This is not accurate in all circumstances, and the OP's situation is an obvious example where it absolutely must be published.

**Reply by Lebowski89** [reply]
> This is not accurate in all circumstances, and the OP's situation is an obvious example where it absolutely must be published.

True. I should have read the initial post better. Apologies. It would be very circumstantial depending on your setup, how you're running it, etc. For my situation, it didn't require any additional changes to my compose file. Just made sure authelia was on same docker network as promethus. In the end, the only change I had to make to get authelia to successfully pass metrics to prometheus was:

```
telemetry:
  metrics:
    enabled: true
    address: 'tcp://:9959/'
```

to

```
telemetry:
  metrics:
    enabled: true
    address: 'tcp://:9959/metrics'
```
And while the prometheus scrape path may be configurable, I'm just going on what the base settings are - working 'out of the box' settings. I've added so many different targets to prometheus the last few days, inc authelia/traefik, radarr/sonarr/lidarr, unRaid, rclone, etc, and all are /metrics scrape path.
